### DISCUSSION TITLE
Not able to add report references to Specflow scenarios, while for NUnit it works fine

### DISCUSSION THREAD

**Discussion by brybski** [allure-framework/allure2#2313] [Questions & Support] [answered]
Hello, I have a project in .NET 6.0 and I'm using:
Specflow [3.9.74]
Nunit [3.13.3]
+ installed Specflow.Allure [3.5.0.73], NUnit.Allure [1.2.1.1] and Allue.Net.Commons [2.11.0].

I followed the instructions on the page and reports generates successfully, however when I start adding references to Specflow scenarios it:
- does not work while using Gherkin tags - it just shows as a label in reports
- throw exception when using AllureApi methods - No test context is active
`System.InvalidOperationException: No test context is active.
   at Allure.Net.Commons.AllureContext.get_CurrentTest()
   at Allure.Net.Commons.AllureLifecycle.UpdateTestCase(Action 1 update)
   at Allure.Net.Commons.AllureApi.SetTestName(String newName)`

What I did for the same project -> I write simple NUnit test and it works perfectly with attributes - it properly displays link to the website.

I have specflow.json and allureConfig.json files ready with Copy to Output directory options marked as Copy always, I tried with different scenarios or putting AllureApi methods in different steps in multiple scenarios but it didn't help. I assume it has no access to TestContext anyhow, but I am not sure how to approach it best.

Did anyone encounter such issue in the past? Thanks in advance.

**Comment by brybski** [comment]
What I also did is trying to replace Specflow.Allure by Allure.Specflow [2.11], but with Specflow.Allure uninstalled (and Allure.Specflow was added as a package] I got such exception:
`OneTimeSetUp: System.ArgumentException : Cannot instantiate implementation type 'Allure.SpecFlowPlugin.AllureBindings' for service type 'Allure.SpecFlowPlugin.AllureBindings'.`

**Comment by delatrie** [comment]
Hi, @brybski !

SpecFlow.Allure, and NUnit.Allure are both deprecated packages and aren't compatible with Allure.Net.Commons 2.11 (which shouldn't be referenced directly, BTW). The only package that you should explicitly add to your SpecFlow project is [Allure.SpecFlow](https://www.nuget.org/packages/Allure.SpecFlow).

I advise you to generate a new project with [Allure Start](https://allurereport.org/start/) and use it as a starting point.

**Reply by brybski** [reply]
Thank you @delatrie, while starting new project through Allure Start it works perfectly fine. However, when I follow these instructions:
[Allure Report Docs - Specflow](https://allurereport.org/docs/specflow/) for existing project, I got exception mentioned above:
`OneTimeSetUp: System.ArgumentException : Cannot instantiate implementation type 'Allure.SpecFlowPlugin.AllureBindings' for service type 'Allure.SpecFlowPlugin.AllureBindings'.`

I tried it on separate branch with Allure.Specflow 2.11 installed. With SpecFlow.Allure it works fine, but as mentioned above - with no option to include AllureApi methods (but it's as expected, since it's not compatible with Allure.Net.Commons 2.11)

I tried to match all files, dependencies and properties from this sample project downloaded from Allure Start but it did not help. Do you have any suspicion what can cause such problem?

**Reply by baev** [reply]
Could you please create a sample repo that reproduces the error?

**Reply by delatrie** [reply]
Upload the faulty project on github and post the link to the repo there.
It would be perfect to also have a GitHub workflow that fails with the error.

**Reply by brybski** [reply]
Sure, thank you guys for quick replies to this problem. I need to prepare sample repo, because the project I'm working on is not private so I cannot share it as a whole. I will do my best to do that till the end of the week.

I was also trying to find a reason by myself and I think it's caused because I'm using [ScenarioDependencies] from SolidToken.Specflow.DependencyInjection. But I will share the sample repo once I have it.

And again, appreciate your support guys! Thanks.

**Comment by brybski** [comment]
@delatrie @baev 
I was able to reproduce it using sample project from here:
https://allurereport.org/start/

Repository is saved here:
https://github.com/brybski/allure-specflow-issue-with-dependencies

What I did:
- added 2 packages: Microsoft.Extensions.DependencyInjection and SolidToken.SpecFlow.DependencyInjection
- created DependenciesHooks class to have ScenarioDependencies attribute

Then exception is thrown when running the test:
`OneTimeSetUp: System.ArgumentException : Cannot instantiate implementation type 'Allure.SpecFlowPlugin.AllureBindings' for service type 'Allure.SpecFlowPlugin.AllureBindings'.`

When I download latest SpecFlow.Allure package it helps and test passes, allure .json files are generated (you can check on separate with-specflow-allure branch in the repo)

**Comment by delatrie** [comment, marked as answer]
`AllureBindings` is a static class that can't be instantiated and shouldn't be managed by a DI container in the first place. I believe that should be fixed on SolidToken's side, but you have some options for working around this.

#### Options 1: disable automatic registration of SpecFlow bindings and add all step definitions by yourself:

That might be tedious, but you can always revert to manual registration of all step definitions used by your project:

```csharp
[Binding]
class DependenciesHooks
{
    [ScenarioDependencies(AutoRegisterBindings = false)]
    public static IServiceCollection CreateServices()
    {
        var serviceCollection = new ServiceCollection();
        serviceCollection.AddScoped<Steps>();
        return serviceCollection;
    }
}
```

#### Option 2: provide your own implementation of IServiceCollection that ignores `AllureBindings`:

```csharp
class AllureIndifferentServiceCollection(ServiceCollection underlyingServiceCollection) : IServiceCollection
{
    readonly ServiceCollection underlyingServiceCollection = underlyingServiceCollection;

    public ServiceDescriptor this[int index]
    {
        get => this.underlyingServiceCollection[index];
        set => this.underlyingServiceCollection[index] = value;
    }

    public int Count => this.underlyingServiceCollection.Count;

    public bool IsReadOnly => this.underlyingServiceCollection.IsReadOnly;

    public void Add(ServiceDescriptor item)
    {
        if (item.ServiceType != typeof(AllureBindings))
        {
            this.underlyingServiceCollection.Add(item);
        }
    }
    public void Clear() => this.underlyingServiceCollection.Clear();
    public bool Contains(ServiceDescriptor item) => this.underlyingServiceCollection.Contains(item);
    public void CopyTo(ServiceDescriptor[] array, int arrayIndex) => this.underlyingServiceCollection.CopyTo(array, arrayIndex);
    public IEnumerator<ServiceDescriptor> GetEnumerator() => this.underlyingServiceCollection.GetEnumerator();
    public int IndexOf(ServiceDescriptor item) => this.underlyingServiceCollection.IndexOf(item);
    public void Insert(int index, ServiceDescriptor item)
    {
        if (item.ServiceType != typeof(AllureBindings))
        {
            this.underlyingServiceCollection.Insert(index, item);
        }
    }
    public bool Remove(ServiceDescriptor item) => this.underlyingServiceCollection.Remove(item);
    public void RemoveAt(int index) => this.underlyingServiceCollection.RemoveAt(index);
    IEnumerator IEnumerable.GetEnumerator() => this.underlyingServiceCollection.GetEnumerator();
}
```

Then, in your `DependenciesHooks.cs`:

```csharp
[Binding]
internal class DependenciesHooks
{
    [ScenarioDependencies]
    public static IServiceCollection CreateServices()
    {
        return new AllureIndifferentServiceCollection([]);
    }
}
```

Both options should fix the issue.

**Reply by brybski** [reply]
Thank you very much, second option helped!
